src/shell.ts

```diff
@@ -0,0 +1,170 @@
+
+    map: T7Map
+    t: Terminal
+    onKey: (ev: KeyboardEvent) => void
+    active = false
+    activeForm: Form | null
+    commands: Map<string, Command>
+    field = ''
+
+    constructor(map: T7Map) {
+        this.map = map
+        this.t = map.t0
+    }
+
+    start() {
+        if (this.active)
+            return
+        this.active = true
+        this.commands = loadCommands(this)
+        let field = ''
+        this.t.scrollToBottom()
+        this.onKey = async ev => {
+            const key = ev.key
+            switch (key) {
+                case "Enter":
+                    this.t.write("\n")
+                    await this.handleLine(field)
+                    field = ''
+                    break
+                case "Backspace":
+                    if (field.length > 0) {
+                        field = field.slice(0, -1)
+                        this.t.write("\b \b")
+                    }
+                    break
+                default:
+                    if (key.length == 1) { // make sure the key is a char
+                        field += key
+                        this.t.write(key)
+                    }
+            }
+            this.field = field
+        }
+    }
+
+    async handleLine(input: string) {
+        const [cmd, ...args] = input.trim().split(/\s+/)
+        await this.execute(cmd, args)
+        this.field = ''
+        this.printPrompt()
+    }
+
+    async execute(cmd: string, args: string[]) {
+        if (this.activeForm) 
+            this.escapeActiveForm()
+        if (!cmd)
+            return
+        const command = this.commands.get(cmd)
+        if (!command)
+            return this.t.writeln(`Command not found: ${cmd}`)
+        this.active = false
+        await command.execute(args)
+        this.active = true
+    }
+
+    async runCommand(cmd: string, args: string[]) {
+        this.escapeActiveForm()
+        this.map.interruptTTY()
+        this.field = [cmd, ...args].join(' ')
+        this.printPrompt()
+        this.t.write("\n")
+        await this.execute(cmd, args)
+        this.field = ''
+        this.printPrompt()
+    }
+
+    async newForm(fields: Fields, type: "menu" | "choice" | "text", title="") {
+        this.escapeActiveForm()
+        this.map.showLog(true)
+        this.t.write("\r\x1B[K")
+        this.t.scrollToBottom()
+        if (title)
+            this.t.writeln(title)
+        this.activeForm = new Form(fields)
+        let promise
+        switch (type) {
+            case "menu":
+                promise = this.activeForm.menu(this.t)
+                break
+            case "choice":
+                promise = this.activeForm.chooseFields(this.t)
+                break
+            case "text":
+                promise = this.activeForm.start(this.t)
+                break
+            default:
+                throw new Error("Unknown form type: " + type)
+        }
+        try {
+            const res = await promise
```

Comment:
While this probably works, it looks strange. Usually I store the function itself into an argument and activate it outside the switch, so this line should be something like:
```
const res = await run(this.t)
```